File: src/main/scala/s/template_method/AbstractDisplay.scala
/* (C) 2022 */
package s.template_method

abstract class AbstractDisplay {
  def open(): Unit

  def print(): Unit

  def close(): Unit

  final def display(): Unit = {
    open()
    for (i <- 0 until 5) {
      print()
    }
    close()
  }
}

File: src/main/kotlin/k/template_method/inheritance/AbstractDisplay.kt
/* (C)2022 */
package k.template_method.inheritance

abstract class AbstractDisplay {
  abstract fun open()

  abstract fun print()

  abstract fun close()

  fun display() {
    open()
    for (i in 0..4) {
      print()
    }
    close()
  }
}
